policykit/policyengine/models.py

```diff
@@ -1026,6 +1027,143 @@ def __init__(self, *args, **kwargs):
+        If false, then we can recover the values from the filters in the following function.
+    """
+        
+    
+    def recover_specs(self):
+        '''
+        In the UI, we will first filter out all ActionFilter instances where the filter_name is not blank and 
+            then show them in the CustomAction tab; 
+        Therefore, we should also be able to recover from the filter function the specified values for this action type
+        '''
+        pass
+
+    def generate_filter(self, specs):
+        '''
+        help generate the filters based on the specifications users make in the frontend
+        '''
+        pass
+    
+    @property
+    def permissions(self):
+        if self.community_name:
+            # If it is a user custom action, it has a new permission name
+            permissions = ((f"can_execute_{self.community_name}", "Can execute {self.community_name}"))
+        else:
+            # Otherwise, this permission for this new CustomAction is the same as the GovernableAction it builts upon 
+            action_content_type = ContentType.objects.filter(model=self.action_type)     
+            all_permissions = Permission.objects.filter(content_type__in=action_content_type)
+            # Search for all permissions related to this GovernableAction
+            permissions = [(perm.codename, perm.name) for perm in all_permissions if perm.codename.startswith("can_execute") ]
+            
+            # While it is obvious that the permission codename is f"can_execute_{self.action_type}", 
+            # We actually do not know exactly the corresponding permisson name
+            # That is why we take such trouble to extract it
+        return permissions
+
+
+    def save(self, *args, **kwargs):
+        """
+        Add the permission if it is a user custom action
+        """
+
+        if not self.pk and self.community_name:
+            action_content_type = ContentType.objects.get_for_model(CustomAction)
+            all_permissions = self.permissions
+            for perm in all_permissions:
+                # TODO not sure whether we should use the content type of CustomAction here
+                Permission.objects.create(codename=perm[0], name=perm[1], content_type=action_content_type)
+                # TODO not sure we would like to assign this permission to all users by default or not
+                # perhaps we should at first assign it to users who have the execute permission of the referenced GovernableAction
+        super(CustomAction, self).save(*args, **kwargs)
+
+class Procedure(models.Model):
+
+    name = models.TextField(blank=True, default='')
+    """the name of the procedure if exists"""
+
+    initial_code = models.TextField(blank=True, default='')
+
+    notify_code = models.TextField(blank=True, default='')
+
+    check_code = models.TextField(blank=True, default='')
+
+    variables_dict = models.TextField(blank=True, default='')
```

Comment:
I'm confused why we're storing this in a dict instead of just directly having a one-to-many to the PolicyVariables table?

Reply:
Alright

Reply:
Actually - if we are to have a PolicyTemplate class, why not store policy variables in there instead of in here? What if you want to use a policy variable in your filter step or execute step?

Reply:
Actually, that's not a good idea since we want Procedures to be able to exist and have PolicyVariables independent of any PolicyTemplate. Hmmmm so I guess we store PolicyVariables in Procedure for now and leave it at that.

Reply:
Yeah, I agree, since we still would like to keep the Procedure class an independent class and the sharing and creation of Procedures also go with their corresponding PolicyVariables